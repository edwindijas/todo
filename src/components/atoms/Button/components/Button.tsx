interface ButtonProps {
  children: React.ReactElement | string,
  className: string,
  onClick?: () => void,
  dataTestId: string
}

 const Button = ({
  className,
  children,
  onClick,
  dataTestId
}: ButtonProps): JSX.Element =>
  <button data-testid={dataTestId} onClick={onClick} className={className}>
    {children}
  </button>


export default Button