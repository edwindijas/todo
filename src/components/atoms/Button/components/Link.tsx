import {useCallback} from 'react'
import { cva } from 'class-variance-authority'
import { Link, NavLink } from 'react-router-dom'

interface CustomLinkProps {
  isNavLink: boolean,
  children: React.ReactElement | string ,
  className: string,
  to: string,
  onClick?: () => void
}

const customLinkVariants = cva('', {
  variants: {
    isActive: {
      true: ['bg-blue-light text-blue-medium fill-blue-medium relative',
      'before:absolute before:top-2 before:bottom-2 before:left-0 before:bg-blue-medium before:w-1 before:block before:rounded-sm'
      ],
      false: 'text-blue-desaturated fill-blue-desaturated'
    }
  },
  defaultVariants: {
    isActive: false
  }
})

const CustomLink = ({
  children,
  className,
  isNavLink,
  to,
  onClick
}: CustomLinkProps): JSX.Element => {

  const getClassName = useCallback(({isActive}: {isActive: boolean}): string => {
    const cvaClass = customLinkVariants({isActive})
    return `${className} ${cvaClass}`
  }, [className])

  if (isNavLink) {
    return<NavLink to={to} className={getClassName} >{children}</NavLink>
  }

  return <Link to={to} className={className} onClick={onClick} >{children}</Link>
}




export default CustomLink 