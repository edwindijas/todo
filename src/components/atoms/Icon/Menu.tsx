import React from 'react'
import ReactComponent from '@material-design-icons/svg/round/menu.svg?react'
import Icon from '.'

const MenuIcon = (props: React.SVGAttributes<SVGElement>): JSX.Element => (
  <Icon {...props} component={ReactComponent} />
)

export default MenuIcon
