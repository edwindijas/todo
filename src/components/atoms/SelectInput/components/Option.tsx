import { cva } from 'class-variance-authority'
import { useMemo } from 'react'

const SelectOption = ({item, selected}: SelectOptionProps): JSX.Element => {

  const isSelected = useMemo<boolean>((): boolean => {

  if (typeof item === 'string' && typeof selected === 'string') {
    return item === selected
  }
  
    return false
  }, [selected])

  const label = typeof item === 'string' ? item : item.value
  return <div className={selectOptionVariants({selected: isSelected})} >
    { label }
  </div>
}


const selectOptionVariants = cva('', {
  variants: {
    selected: {
      true: 'bg-blue-medium text-white'
    }
  }
})


interface SelectOptionProps {
  item: SelectOptionItem
  selected: SelectOptionItem
  onClick?: (v: SelectOptionItem) => void
}

export type SelectOptionItem = string | {
  value: string
  id: string
}

export default SelectOption