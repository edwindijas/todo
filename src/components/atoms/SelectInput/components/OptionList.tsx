import { useState } from 'react'
import SelectOption, { type SelectOptionItem } from './Option'

const OptionsList = (
  {
    options
  }: OptionListProps
): JSX.Element => {
  const [selected, setSelected] = useState<SelectOptionItem>('')
  
  return  <ul >
  {
    options.map((option) => <li role='select-option' key={typeof option === 'string' ? option : option.id} >
      <SelectOption onClick={setSelected} selected={selected} item={option} />
    </li>)
  }
</ul>
}

interface OptionListProps {
  options: SelectOptionItem[]
}

export default OptionsList