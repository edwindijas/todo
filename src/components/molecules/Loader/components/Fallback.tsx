const Fallback = (): JSX.Element => (
  <div className='w-full h-full flex items-center justify-center'>
    Loading...
  </div>
)
export default Fallback