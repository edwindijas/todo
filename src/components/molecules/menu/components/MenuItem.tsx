import Button from '@/components/atoms/Button'

export interface MenuItemProps {
  startIcon?: React.ReactElement,
  endIcon?: React.ReactElement,
  label: string,
  to: string
  isNavLink?: boolean
}

const MenuItem = ({
  startIcon,
  endIcon,
  label,
  to,
  isNavLink,
}: MenuItemProps): JSX.Element =>
 <Button
  startIcon={startIcon}
  isNavLink={isNavLink}
  isLink to={to}
  endIcon={endIcon}
  padding='lg'
>{ label }</Button>

export default MenuItem
