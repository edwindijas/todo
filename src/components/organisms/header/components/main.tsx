import Button from '@/components/atoms/Button'
import AddIcon from '@/components/atoms/Icon/Add'
import CalendarMonthIcon from '@/components/atoms/Icon/CalendarMonth'
import DashboardOutlineIcon from '@/components/atoms/Icon/DashboardOutline'
import GroupsOutlineIcon from '@/components/atoms/Icon/GroupsOutline'
import LogoIcon from '@/components/atoms/Icon/Logo'
import SettingsOutlineIcon from '@/components/atoms/Icon/SettingsOutline'
import TaskAltIcon from '@/components/atoms/Icon/TaskAlt'
import ViewWeekOutlineIcon from '@/components/atoms/Icon/ViewWeekOutline'
import Menu from '@/components/molecules/menu'
import { type MenuItemProps } from '@/components/molecules/menu/components/MenuItem'

const menuItems: MenuItemProps[] = [
  {
    label: 'dashboard',
    to: '/',
    startIcon: <DashboardOutlineIcon />
  },
  {
    label: 'tasks',
    to: '/tasks',
    startIcon: <TaskAltIcon />
  },
  {
    label: 'calendar',
    to: '/calendar',
    startIcon: <CalendarMonthIcon />
  },
  {
    label: 'board',
    to: '/board',
    startIcon: <ViewWeekOutlineIcon />
  },
  {
    label: 'project-settings',
    to: '/project-settings',
    startIcon: <SettingsOutlineIcon />
  },
  {
    label: 'members',
    to: '/members',
    startIcon: <GroupsOutlineIcon />
  }
]

const Main = (): JSX.Element => <div className='w-64 h-full bg-white relative'>
  <div >
    <div className='flex py-4 px-6 gap-2 items-center'>
      <div className='w-8 h-8'><LogoIcon className=' fill-blue-dark' /></div>
      <p className='text-md text-blue-dark flex flex-col gap-0 leading-4'>
        <span>TASK</span>
        <span className='font-extrabold'>SPACE</span>
      </p>
    </div>
    <div>
    </div>
    <div className='p-4' >
      <Menu items={menuItems} isNavLink />
    </div>
  </div>
  <div className='px-6 absolute bottom-4 w-full'>
    <Button 
      isLink 
      intent='secondary'
      className='w-full justify-center'
      startIcon={<AddIcon />}
      float
      >New Tasks</Button>
  </div>
</div>

export default Main