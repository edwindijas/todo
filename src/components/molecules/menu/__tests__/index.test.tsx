import Menu  from '..';
import render from '@/utils/test-utils/customRender';
import { type MenuItemProps } from '../components/MenuItem';

const dummyMenu: Array<MenuItemProps | string> = [
  'one',
  'two',
  'three',
  {
    label: 'Hello',
    to: 'Bingo',
    startIcon: <p>Start</p>,
    endIcon: <p>Start</p>
  }
]

describe ('Menu', () => {
  it ('matches snapnshot', () => {
    const {container} = render(<>
      <Menu items={dummyMenu} />
      <Menu items={dummyMenu} isNavLink />
    </>)
    expect(container).toMatchSnapshot()
  })
})