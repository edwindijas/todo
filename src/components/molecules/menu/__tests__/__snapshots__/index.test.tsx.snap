// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu matches snapnshot 1`] = `
<div>
  <ul
    class="flex flex-col"
  >
    <li>
      <a
        class=" flex capitalize items-center gap-2 relative px-4 h-12 rounded-md"
        href="/one"
      >
        <span
          class="w-full"
        >
          one
        </span>
      </a>
    </li>
    <li>
      <a
        class=" flex capitalize items-center gap-2 relative px-4 h-12 rounded-md"
        href="/two"
      >
        <span
          class="w-full"
        >
          two
        </span>
      </a>
    </li>
    <li>
      <a
        class=" flex capitalize items-center gap-2 relative px-4 h-12 rounded-md"
        href="/three"
      >
        <span
          class="w-full"
        >
          three
        </span>
      </a>
    </li>
    <li>
      <a
        class=" flex capitalize items-center gap-2 relative px-4 h-12 rounded-md"
        href="/Bingo"
      >
        <div
          class="shrink-0 h-6 w-6"
        >
          <p>
            Start
          </p>
        </div>
        <span
          class="w-full"
        >
          Hello
        </span>
        <div>
          <p>
            Start
          </p>
        </div>
      </a>
    </li>
  </ul>
  <ul
    class="flex flex-col"
  >
    <li>
      <a
        class=" flex capitalize items-center gap-2 relative px-4 h-12 rounded-md text-blue-desaturated fill-blue-desaturated"
        href="/one"
      >
        <span
          class="w-full"
        >
          one
        </span>
      </a>
    </li>
    <li>
      <a
        class=" flex capitalize items-center gap-2 relative px-4 h-12 rounded-md text-blue-desaturated fill-blue-desaturated"
        href="/two"
      >
        <span
          class="w-full"
        >
          two
        </span>
      </a>
    </li>
    <li>
      <a
        class=" flex capitalize items-center gap-2 relative px-4 h-12 rounded-md text-blue-desaturated fill-blue-desaturated"
        href="/three"
      >
        <span
          class="w-full"
        >
          three
        </span>
      </a>
    </li>
    <li>
      <a
        class=" flex capitalize items-center gap-2 relative px-4 h-12 rounded-md text-blue-desaturated fill-blue-desaturated"
        href="/Bingo"
      >
        <div
          class="shrink-0 h-6 w-6"
        >
          <p>
            Start
          </p>
        </div>
        <span
          class="w-full"
        >
          Hello
        </span>
        <div>
          <p>
            Start
          </p>
        </div>
      </a>
    </li>
  </ul>
</div>
`;
